packages/flutter/lib/src/gestures/force_press.dart

```diff
@@ -0,0 +1,260 @@
+  /// argument, which is a [ForcePressDetails] object.
+  GestureForcePressPeakCallback onPeak;
+
+  /// A pointer is no longer in contact with the screen.
+  ///
+  /// The position of the pointer is provided in the callback's `details`
+  /// argument, which is a [ForcePressDetails] object.
+  GestureForcePressEndCallback onEnd;
+
+  /// The pressure of the press required to initiate a force press.
+  ///
+  /// A value of 0.0 is no pressure, and 1.0 is maximum pressure.
+  double startPressure;
+
+  /// The pressure of the press required to peak a force press.
+  ///
+  /// A value of 0.0 is no pressure, and 1.0 is maximum pressure.
+  double peakPressure;
+
+  Duration _lastTimeStamp;
+  Offset _lastPosition;
+  _ForceState _state = _ForceState.ready;
+
+  @override
+  void addPointer(PointerEvent event) {
+    startTrackingPointer(event.pointer);
+    if (_state == _ForceState.ready) {
+      _state = _ForceState.possible;
+      _lastPosition = event.position;
+    }
+  }
+
+  @override
+  void handleEvent(PointerEvent event) {
+    assert(_state != _ForceState.ready);
+    // A static pointer with changes in pressure creates PointerMoveEvent events.
+    if (event is PointerMoveEvent || event is PointerDownEvent) {
+      final double pressure = _inverseLerp(event.pressureMin, event.pressureMax, event.pressure);
+      if (_state == _ForceState.possible) {
+        _lastTimeStamp = event.timeStamp;
+        _lastPosition = event.position;
+        if (pressure > startPressure) {
+          _state = _ForceState.started;
+          resolve(GestureDisposition.accepted);
+        } else if (event.delta.distanceSquared > kTouchSlop)
+          resolve(GestureDisposition.rejected);
+      } else if (_state == _ForceState.accepted || _state == _ForceState.peaked || _state == _ForceState.started) {
+        // Two separate if conditionals so that update will be called during the same frame as start.
+        if (pressure > startPressure && _state != _ForceState.started && _state != _ForceState.peaked) {
+          _state = _ForceState.started;
+          _gestureStarted();
+        }
+        if (_state == _ForceState.started || _state == _ForceState.peaked) {
+          if (onPeak != null && pressure > peakPressure && _state != _ForceState.peaked) {
+            _state = _ForceState.peaked;
+            invokeCallback<void>('onPeak', () => onPeak(ForcePressDetails(
+              globalPosition: event.position,
+              sourceTimeStamp: event.timeStamp,
+            )));
+          }
+          if (onUpdate != null) {
+            invokeCallback<void>('onUpdate', () => onUpdate(ForcePressUpdateDetails(
+              sourceTimeStamp: event.timeStamp,
+              pressure: pressure,
+              globalPosition: event.position,
+            )));
+          }
+        }
+      }
+    }
+    stopTrackingIfPointerNoLongerDown(event);
+    _lastTimeStamp = event.timeStamp;
```

Comment:
not clear why this is present twice in this function. Don't you want to do this regardless?